### PR TITLE
Add a derived data set to compute churn

### PR DESCRIPTION
**Pull request by mreid-moz**
**DO NOT MERGE**

Just looking for feedback.  I set the output bucket in `src/main/resources/application.conf` to `telemetry-test-bucket`, and I am seeing the following error when I run this code:

```
<...snip...>
16/01/08 17:19:34 INFO ShuffleBlockFetcherIterator: Started 0 remote fetches in 0 ms
16/01/08 17:19:37 INFO Executor: Finished task 35.0 in stage 4.0 (TID 47). 1165 bytes result sent to driver
16/01/08 17:19:37 INFO TaskSetManager: Starting task 40.0 in stage 4.0 (TID 52, localhost, PROCESS_LOCAL, 2170 bytes)
16/01/08 17:19:37 INFO Executor: Running task 40.0 in stage 4.0 (TID 52)
16/01/08 17:19:37 INFO TaskSetManager: Finished task 35.0 in stage 4.0 (TID 47) in 24141 ms on localhost (37/41)
16/01/08 17:19:37 INFO ShuffleBlockFetcherIterator: Getting 4 non-empty blocks out of 4 blocks
16/01/08 17:19:37 INFO ShuffleBlockFetcherIterator: Started 0 remote fetches in 0 ms
16/01/08 17:19:40 INFO Executor: Finished task 40.0 in stage 4.0 (TID 52). 1165 bytes result sent to driver
16/01/08 17:19:40 INFO TaskSetManager: Finished task 40.0 in stage 4.0 (TID 52) in 3725 ms on localhost (38/41)
16/01/08 17:19:49 INFO Executor: Finished task 38.0 in stage 4.0 (TID 50). 1165 bytes result sent to driver
16/01/08 17:19:49 INFO TaskSetManager: Finished task 38.0 in stage 4.0 (TID 50) in 19835 ms on localhost (39/41)
16/01/08 17:19:54 INFO Executor: Finished task 39.0 in stage 4.0 (TID 51). 1165 bytes result sent to driver
16/01/08 17:19:54 INFO TaskSetManager: Finished task 39.0 in stage 4.0 (TID 51) in 20263 ms on localhost (40/41)
^R
16/01/08 17:20:04 INFO Executor: Finished task 36.0 in stage 4.0 (TID 48). 1165 bytes result sent to driver
16/01/08 17:20:04 INFO TaskSetManager: Finished task 36.0 in stage 4.0 (TID 48) in 48467 ms on localhost (41/41)
16/01/08 17:20:04 INFO DAGScheduler: ResultStage 4 (foreach at SimpleDerivedStream.scala:34) finished in 123.510 s
16/01/08 17:20:04 INFO TaskSchedulerImpl: Removed TaskSet 4.0, whose tasks have all completed, from pool
16/01/08 17:20:04 INFO DAGScheduler: Job 1 finished: foreach at SimpleDerivedStream.scala:34, took 125.296149 s
16/01/08 17:20:04 ERROR ContextCleaner: Error in cleaning thread
java.lang.InterruptedException
	at java.lang.Object.wait(Native Method)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:135)
	at org.apache.spark.ContextCleaner$$anonfun$org$apache$spark$ContextCleaner$$keepCleaning$1.apply$mcV$sp(ContextCleaner.scala:157)
	at org.apache.spark.util.Utils$.tryOrStopSparkContext(Utils.scala:1136)
	at org.apache.spark.ContextCleaner.org$apache$spark$ContextCleaner$$keepCleaning(ContextCleaner.scala:154)
	at org.apache.spark.ContextCleaner$$anon$3.run(ContextCleaner.scala:67)
16/01/08 17:20:04 ERROR Utils: uncaught error in thread SparkListenerBus, stopping SparkContext
java.lang.InterruptedException
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:996)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1303)
	at java.util.concurrent.Semaphore.acquire(Semaphore.java:317)
	at org.apache.spark.util.AsynchronousListenerBus$$anon$1$$anonfun$run$1.apply$mcV$sp(AsynchronousListenerBus.scala:65)
	at org.apache.spark.util.Utils$.tryOrStopSparkContext(Utils.scala:1136)
	at org.apache.spark.util.AsynchronousListenerBus$$anon$1.run(AsynchronousListenerBus.scala:63)
16/01/08 17:20:05 INFO BlockManagerInfo: Removed broadcast_3_piece0 on localhost:61555 in memory (size: 35.0 KB, free: 530.2 MB)
16/01/08 17:20:05 INFO ContextCleaner: Cleaned accumulator 4
16/01/08 17:20:05 INFO BlockManagerInfo: Removed broadcast_2_piece0 on localhost:61555 in memory (size: 36.7 KB, free: 530.2 MB)
16/01/08 17:20:05 INFO ContextCleaner: Cleaned accumulator 3
16/01/08 17:20:05 INFO ContextCleaner: Cleaned shuffle 1
16/01/08 17:20:05 INFO BlockManagerInfo: Removed broadcast_1_piece0 on localhost:61555 in memory (size: 36.5 KB, free: 530.2 MB)
16/01/08 17:20:05 INFO ContextCleaner: Cleaned accumulator 2
16/01/08 17:20:05 INFO BlockManagerInfo: Removed broadcast_0_piece0 on localhost:61555 in memory (size: 35.8 KB, free: 530.3 MB)
16/01/08 17:20:05 INFO ContextCleaner: Cleaned accumulator 1
16/01/08 17:20:05 INFO ContextCleaner: Cleaned shuffle 0
16/01/08 17:20:05 INFO SparkUI: Stopped Spark web UI at http://192.168.2.12:4040
16/01/08 17:20:05 INFO DAGScheduler: Stopping DAGScheduler
[success] Total time: 150 s, completed 8-Jan-2016 5:20:05 PM
16/01/08 17:20:05 INFO DiskBlockManager: Shutdown hook called
16/01/08 17:20:05 INFO ShutdownHookManager: Shutdown hook called
16/01/08 17:20:05 INFO ShutdownHookManager: Deleting directory /private/var/folders/11/11jyx2zd2k75487c1y762hwr0000gn/T/spark-522d2ca9-f1a0-4f8a-a428-2b87e69d0280/userFiles-e835b8ed-9b30-4e39-82fc-3ddf8c514c3d
16/01/08 17:20:05 INFO ShutdownHookManager: Deleting directory /private/var/folders/11/11jyx2zd2k75487c1y762hwr0000gn/T/spark-522d2ca9-f1a0-4f8a-a428-2b87e69d0280
```

I don't see anything appear in the output bucket, any suggestions?

### CODE REVIEW

**Comment by mreid-moz** [comment]
My run command was:
`sbt "run-main telemetry.DerivedStream --from-date 20160107 --to-date 20160107 Churn"`

**Comment by vitillo** [comment]
You can safely [disregard](https://mail-archives.apache.org/mod_mbox/incubator-spark-user/201406.mbox/%3CCAMJOb8=6Q9xDhLALyo-bWQR3AkMz=ost=7jhTad7Zduv7mpQ0w@mail.gmail.com%3E) the error that appears at the very end of the excution.

**Comment by mreid-moz** [comment]
OK, I think this is ready for review now.

File: src/main/scala/streams/Churn.scala [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+      case x: JInt => x.num.toInt > 0
+      case _ => false
+    }
+  }
+
+  // Given histogram h, return true if it has a value in the "true" bucket,
+  // or false if it has a value in the "false" bucket, or None otherwise.
+  def booleanHistogramToBoolean(h: JValue): Option[Boolean] = {
+    var result: Option[Boolean] = None
+    if (gtZero(h \ "values" \ "1")) {
```
Comment (vitillo) [inline src/main/scala/streams/Churn.scala]: You should pattern match on both histogram values at the same time as it makes it more readable.

File: src/main/scala/streams/Churn.scala [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+      .set("syncCountDesktop", weaveDesktop match {
+             case Some(x) => x
+             case _ => null
+           })
+      .set("syncCountMobile", weaveMobile match {
+             case Some(x) => x
+             case _ => null
+           })
+      .build
+//    println("Matched one record")
```
Comment (vitillo) [inline src/main/scala/streams/Churn.scala]: Please remove commented code.

File: src/main/scala/streams/Churn.scala [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+          }
+        }
+        if (topBucket >= 0) {
+          Some(topBucket)
+        } else {
+          None
+        }
+      }
+      case _ => {
+//        println("Unexpected format for enum histogram")
```
Comment (vitillo) [inline src/main/scala/streams/Churn.scala]: Please remove commented code.

File: src/test/scala/ChurnTest.scala [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+//    println("---------")
+//    println("payload:")
+//    println(compact(render(json \ "payload")))
+//    println("---------")
+//    println("histograms:")
+//    println(compact(render(histograms)))
+//    println("---------")
+//    println("tb1:")
+//    println(compact(render(histograms \ "TEST_BOOLEAN_1")))
+//    println("---------")
```
Comment (vitillo) [inline src/test/scala/ChurnTest.scala]: Please remove commented code.

File: src/main/scala/streams/Churn.scala [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+  // Find the largest numeric bucket that contains a value greater than zero.
+  def enumHistogramToCount(h: JValue): Option[Long] = {
+    var result: Option[Long] = (h \ "values") match {
+      case JNothing => None
+      case JObject(x) => {
+        var topBucket = -1
+        for ((k, v) <- x) {
+          val bucket = toInt(k).getOrElse(-1)
+          if (bucket > topBucket) {
+            if (gtZero(v)) {
```
Comment (vitillo) [inline src/main/scala/streams/Churn.scala]: There is no need to nest the extraction of optional values. Have a look at this [blog post](http://danielwestheide.com/blog/2012/12/19/the-neophytes-guide-to-scala-part-5-the-option-type.html) under "for comprehensions".

File: src/main/scala/streams/Churn.scala [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+  }
+  
+  // Find the largest numeric bucket that contains a value greater than zero.
+  def enumHistogramToCount(h: JValue): Option[Long] = {
+    var result: Option[Long] = (h \ "values") match {
+      case JNothing => None
+      case JObject(x) => {
+        var topBucket = -1
+        for ((k, v) <- x) {
+          val bucket = toInt(k).getOrElse(-1)
```
Comment (vitillo) [inline src/main/scala/streams/Churn.scala]: No need to nest the extraction of optional values, see below.

**Comment by vitillo** [comment]
Looks good.

**Pull request merged**